db/compaction_iterator.cc

```diff
@@ -269,6 +279,20 @@ void CompactionIterator::NextFromInput() {
     // compaction filter). ikey_.user_key is pointing to the copy.
     if (!has_current_user_key_ ||
         !cmp_->Equal(ikey_.user_key, current_user_key_)) {
+      key_cnt_++;
+      // Use key_cnt_ to reduce the overhead of reading current time
+      if (snap_list_callback_ && snapshots_->size() &&
+          key_cnt_ % snap_list_callback_->key_skip() == 0) {
+        const uint64_t nanos = snap_list_callback_->snap_refresh_nanos();
```

Comment:
I feel that neither skip_skip() and snap_refresh_nanos should belong to callback. Can we make it a variable of Compaction object instead?

Reply:
I am inclined towards keeping them here. The more variables a class has, the harder it is to read the code and get a grasp of what it does. These two variables will not be used unless snap_list_callback_ is involved so the complexity that comes with them could be localized in snap_list_callback_.

Reply:
Encapsulating is one major goal of class. The code of using key_skip_ and snap_refresh_nanos_ are only in CompactionIterator, and no function inside SnapListCallback is supposed to use them. To me it doesn't belong to SnapListCallback.
If you create a function in SnapListCallback and does the calling check, then it feels that these two variables should resident there, which might not be a bad idea.

Reply:
Makes sense.
The latest patch had already removed accessing key_skip_ from outside SnapListCallback. Let me see if I can do the same for snap_refresh_nanos_. Otherwise will bring it out in CompactionIterator.